src/material_library.cpp

```diff
@@ -0,0 +1,221 @@
+  // fastest-way-to-check-if-a-file-exist-using-standard-c-c11-c
+  std::ifstream infile(filename.c_str());
+  return infile.good();
+}
+
+// Append Material to the library from hdf5 file
+void MaterialLibrary::from_hdf5(std::string filename, std::string datapath,
+                                int protocol) {
+  const char* data_path = datapath.c_str();
+
+  if (!hdf5_path_exists(filename, data_path)) return;
+
+  int num_materials = get_length_of_table(filename, datapath);
+
+  for (int i = 0; i < num_materials; i++) {
+    pyne::Material mat;  // from file
+    // read new mat
+    mat.from_hdf5(filename, datapath, i, protocol);
+
+    append_to_nuclist(mat);
+    material_library[mat.metadata["name"].asString()] = mat;
+    matlist.insert(mat.metadata["name"].asString());
+  }
+}
+
+void MaterialLibrary::add_material(pyne::Material mat) {
+  std::string mat_name = mat.metadata["name"].asString();
+  std::pair<pyne::mat_iter, bool> ins = material_library.insert(
+      std::pair<std::string, pyne::Material>(mat_name, mat));
+
+  if (ins.second) {
+    append_to_nuclist(mat);
+    matlist.insert(mat.metadata["name"].asString());
+  }
+}
+
+void MaterialLibrary::del_material(pyne::Material mat) {
+  std::string mat_name = mat.metadata["name"].asString();
+  del_material(mat_name);
+}
+
+void MaterialLibrary::del_material(std::string mat_name) {
+  material_library.erase(mat_name);
+  matlist.erase(mat_name);
+}
+
+pyne::Material MaterialLibrary::get_material(std::string mat_name) {
+  pyne::mat_iter it = material_library.find(mat_name);
+  if (it != material_library.end()) {
+    return it->second;
+  } else {
+    return pyne::Material();
+  }
+}
+
+void MaterialLibrary::write_hdf5(std::string filename, std::string datapath,
+                                 std::string nucpath, int chunksize) {
+  // Turn off annoying HDF5 errors
+  H5Eset_auto2(H5E_DEFAULT, NULL, NULL);
+
+  // Set file access properties so it closes cleanly
+  hid_t fapl;
+  fapl = H5Pcreate(H5P_FILE_ACCESS);
+  H5Pset_fclose_degree(fapl, H5F_CLOSE_STRONG);
+  // Create new/open datafile.
+  hid_t db;
+  if (pyne::file_exists(filename)) {
+    bool ish5 = H5Fis_hdf5(filename.c_str());
+    if (!ish5) throw h5wrap::FileNotHDF5(filename);
+    db = H5Fopen(filename.c_str(), H5F_ACC_RDWR, fapl);
+  } else
+    db = H5Fcreate(filename.c_str(), H5F_ACC_TRUNC, H5P_DEFAULT, fapl);
+
+  pyne::mat_map material_library_tmp = material_library;
+  std::set<int> nuclist_tmp = nuclist;
+  // We are using material_library and nuclist copy to allow adding exisitng
+  // nuclide/materials from the file
+
+  std::vector<int> nuclides;
+  nuclides.clear();
+  std::copy(nuclist_tmp.begin(), nuclist_tmp.end(),
+            inserter(nuclides, nuclides.begin()));
+
+  int nuc_size;
+  hsize_t nuc_dims[1];
+  nuc_size = nuclides.size();
+
+  // Create the data if it doesn't exist
+  int nuc_data[nuc_size];
+  int n = 0;
+  for (auto it = nuclides.begin(); it != nuclides.end(); it++) {
```

Comment:
You should be able to shorten this to `for (auto nuclide : nuclides)` here.

Reply:
ok I am using auto, anyway, thanks for the `auto` suggestion this is a good think to know!